Update image_path() invocations to include version arguments

jmarshall opened the pull request:
Add version arguments corresponding to each current latest image in our Artifact Registry. Where a source file uses an image repeatedly, abstract into a `FOO_VERSION` variable per-source-file.

(Our repository has no `bwamem2` image. Use 2.2.1 which was the current version when the relevant _align.py_ code was added and which still is.)

There are a few more instances still to do, mostly for **cpg_workflows** which is nontrivial as the desired version is probably “today's version” which moves regularly. Is there an equivalent in the cpg_flow new world order? If not, perhaps we can just leave these ones as single-argument `image_path()`.

This PR uses the `VERSION-1`, `VERSION-2`, etc style tags as per the images revamp, with a suffix indicating changes in CPG's packaging of the tool. Additional tags for many images in the artifact registry were added on June 4th by running this [_create-extra-tags.sh_ script](https://github.com/user-attachments/files/22035202/create-extra-tags.sh.txt), which added initial `x.y-1` tags to existing images using commands such as

```
gcloud artifacts docker tags add    \
    australia-southeast1-docker.pkg.dev/cpg-common/images/bcftools:1.16    \
    australia-southeast1-docker.pkg.dev/cpg-common/images/bcftools:1.16-1
```

Using these `x.y-1` tags pulls exactly the same images as the existing `x.y` tags.

jmarshall committed:
Update image_path() invocations to include version arguments

Add version arguments corresponding to each current latest image in our
Artifact Registry. Where a source file uses an image repeatedly, abstract
into a FOO_VERSION variable per-source-file.

(Our repository has no bwamem2 image. Use 2.2.1 which was the current
version when the relevant align.py code was added and which still is.)

Require at least cpg-utils 5.3.0, which has 2-argument image_path().

vivbak commented:
If images is backward compatible with no specification for now, then I think leaving cpg_workflows as the exception is all good. The updates on cpg_flow will be less frequent, and indeed should have users adopting them as a choice. CC @violetbrina @nevoodoo

jmarshall committed:
Need to authenticate to GCP to use image_path() with a version

jmarshall committed:
Don't test bwa-mem2 as we have no image for it

While we're here, fix "alinger" typo.

jmarshall committed:
Test that image versions specified in workflows actually exist

The workflows these are testing now use `image_path('somalier', 'version')`
which queries the Artifact Registry to get the actual image URL. Update the
tests to check that the requested image version does exist rather than the
existing very basic image tests. Adjust images['somalier'] et al so that
the assert still succeeds even if a workflow uses `image_path('somalier')`
without a version argument.

jmarshall committed:
Update bcftools 1.20 and 1.21 image_path() invocations

Switch these from bespoke bcftools_120 and bcftools_121 images to use
the new 1.20-1 and 1.21-1 tagged versions of the main bcftools image.

jmarshall committed:
Update VEP image_path() invocations

Because vep_version is used in the mount path and to switch between
105-style and 110-style behaviours, leave it as is and introduce an
additional vep_image config entry. We could have computed vep_version
as the "basename" of vep_image, but that would be problematic if we
eventually deploy e.g. 111.1-1 that also behaves like 110.

jmarshall commented on cpg_workflows/jobs/gcnv.py:
```diff
@@ -16,6 +16,8 @@
 from cpg_workflows.scripts import upgrade_ped_with_inferred
 from cpg_workflows.utils import can_reuse, chunks
 
+GATK_VERSION = '4.5.0.0-1'
```
That reminds me that I also punted on figuring out what to do about _gatk_sv_common.py_'s `get_images()`…

jmarshall committed:
Fix image_path('gatk_gcnv'), which was previously overridden by config

Update GATK images used in the gCNV pipeline to the sv/gatk image
actually used, as specified in the now-unnecessary config override.

jmarshall commented:
I have now verified that all the 2-argument `image_path()` calls in this PR do indeed point to images that exist. With one exception, all the remaining `[images]` overrides in _configs/defaults/*.toml_ are for use with `get_images()` and exist outside our image-handling system — in fact, `get_images()` could perhaps be refactored to just use `config_retrieve()` directly.

The one exception is `ica` in _configs/defaults/realign_genotype_with_dragen.toml_, which appears to be unused.

Alexander-Stuckey commented:
> The one exception is `ica` in _configs/defaults/realign_genotype_with_dragen.toml_, which appears to be unused.

The `ica` image is no longer needed. Dragen has migrated to it's own repo, and builds it's own image there (`dragen_align_pa`)

Alexander-Stuckey commented:
All good with PA @jmarshall

MattWellie commented:
I understand the intention, I'm pointing out that it hasn't been tested in practice, and alters every workflow we run. There's no way we would normally accept an untested change of this scale under time pressure.
